src/deploy/AllocatorInit.sol

```diff
@@ -0,0 +1,141 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.8.0;
+
+import { DssInstance } from "dss-test/MCD.sol";
+import { AllocatorSharedInstance, AllocatorCoreInstance, AllocatorFunnelsInstance } from "./AllocatorInstances.sol";
+
+interface RolesLike {
+    function setIlkAdmin(bytes32, address) external;
+    function setUserRole(bytes32, address, uint8, bool) external;
+    function setRoleAction(bytes32, uint8, address, bytes4, bool) external;
+}
+
+interface RegistryLike {
+    function file(bytes32, bytes32, address) external;
+}
+
+interface VaultLike {
+    function ilk() external view returns (bytes32);
+    function file(bytes32, address) external;
+    function init() external;
+    function draw(uint256) external;
+    function wipe(uint256) external;
+}
+
+interface BufferLike {
+    function approve(address, address, uint256) external;
+}
+
+interface SwapperLike {
+    function swap(address, address, uint256, uint256, address, bytes calldata) external;
+}
+
+struct AllocatorConfig {
+    uint256 debtCeiling;
+    address allocatorProxy;
+}
+
+struct CoreConfig {
+    uint8 operatorRole;
+    address facilitator;
+    address nst;
+}
+
+struct FunnelsConfig {
+    uint8 operatorRole;
+    address facilitator;
+    address keeper;
+}
+
+library AllocatorInit {
+    uint256 constant WAD = 10 ** 18;
+    uint256 constant RAY = 10 ** 27;
+    uint256 constant RAD = 10 ** 45;
+
+    // TODO: sanity checks
+    // TODO: add contracts to changelog
+    // TODO: add to ilk registry?
+
+    // Should be called from Pause Proxy
+    function initAllocator(
+        DssInstance memory dss,
+        AllocatorSharedInstance memory sharedInstance,
+        AllocatorCoreInstance memory coreInstance,
+        AllocatorConfig memory config
+    ) internal {
+
+        bytes32 ilk = VaultLike(coreInstance.vault).ilk();
+        dss.vat.init(ilk);
+        dss.vat.slip(ilk, coreInstance.vault, int256(1_000_000 * WAD));
+        dss.jug.init(ilk);
+
+        require(config.debtCeiling < WAD, "AllocatorInit/incorrect-ilk-line-precision");
+        dss.vat.file(ilk, "line", config.debtCeiling * RAD);
+        dss.vat.file("Line", dss.vat.Line() + config.debtCeiling * RAD);
+
+        dss.spotter.file(ilk, "pip", sharedInstance.oracle);
+        dss.spotter.file(ilk, "mat", RAY); // TODO: do we need this?
+        dss.spotter.poke(ilk);
+
+        RolesLike(sharedInstance.roles).setIlkAdmin(ilk, config.allocatorProxy);
+        RegistryLike(sharedInstance.registry).file(ilk, "buffer", coreInstance.buffer);
```

Comment:
With this I'm wondering if it would be better that the pauseProxy can input which is the allocatorproxy for each ilk and then, it is the allocatorproxy the one that can fill in the buffer, as otherwise a buffer migration will require a pause proxy component.

Reply:
Currently a buffer migration would involve the pause proxy anyway since the buffer is immutable in the AllocatorVault.
So you would have to replace the allocator vault (which involves the pause proxy anyway since you need to slip).

I think if we do this change it should be in order to make it similar to the roles contract (as I'm not sure we can justify why they are different, except maybe for the fact that the buffer is not expected to change at all).